### PR TITLE
Don't suppress jsdom error reporting in our tests

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/react/issues/8260.

It removes the `error` event listener which caused all jsdom errors to be ignored. We still *silence* some of the errors, but at `console.error` call time. The benefit of this approach compared to our existing `event.preventDefault()` call is that we still *fail the test* if the error was unexpected (like described in https://github.com/facebook/react/issues/8260). And it's nice to remove the weird hack with mucking the error prototype.

I implemented silencing by adding a few special cases to our `console.error` mocks that ignore certain messages (see `shouldIgnoreConsoleError.js`). I've spent a few days trying several other explicit approaches before that. They were all too annoying and complicated, and led to too much noise in tests themselves. Especially because we have many different configurations (dev, prod, bundles). Ignoring ended up the most straightforward (and hopefully maintainable) solution. And that's what we were effectively doing anyway, except at an earlier stage.

For the few cases where we were testing logging itself, I changed those to directly overwrite `console.error` and assert on the mock calls. Seems easier than remembering what gets mocked and when. I also made the test non-internal while I was at it.

### CODE REVIEW

**Comment by react-linter** [comment]
ReactDOM: size: -0.3%, gzip: -0.3%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 5550ed4a8f33ff562e807175b8f35a1f9e037670...b8f8cf69d1519648ca7a5d8042a54241e81ac40f</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | -0.0% | -0.0% | 643.46 KB | 643.35 KB | 151.44 KB | 151.41 KB | UMD_DEV
**react-dom.production.min.js** | **-0.3%** | **-0.3%** | **96.28 KB** | **96 KB** | **31.21 KB** | **31.12 KB** | **UMD_PROD**
react-dom.development.js | -0.0% | -0.0% | 639.59 KB | 639.49 KB | 150.29 KB | 150.26 KB | NODE_DEV
**react-dom.production.min.js** | **-0.3%** | **-0.3%** | **96.27 KB** | **95.99 KB** | **30.79 KB** | **30.68 KB** | **NODE_PROD**
ReactDOM-dev.js | -0.0% | -0.0% | 646.22 KB | 646.11 KB | 148.36 KB | 148.33 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **-0.0%** | **-0.0%** | **279.01 KB** | **278.92 KB** | **52.28 KB** | **52.26 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | -0.3% | -0.3% | 97.48 KB | 97.2 KB | 31.19 KB | 31.1 KB | NODE_PROFILING
ReactDOM-profiling.js | -0.0% | -0.0% | 281.42 KB | 281.33 KB | 52.92 KB | 52.9 KB | FB_WWW_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | -0.0% | -0.0% | 434.94 KB | 434.83 KB | 98.73 KB | 98.71 KB | UMD_DEV
**react-art.production.min.js** | **-0.3%** | **-0.3%** | **85.16 KB** | **84.88 KB** | **26.4 KB** | **26.33 KB** | **UMD_PROD**
react-art.development.js | -0.0% | -0.0% | 367.46 KB | 367.36 KB | 81.67 KB | 81.64 KB | NODE_DEV
**react-art.production.min.js** | **-0.6%** | **-0.7%** | **50.15 KB** | **49.87 KB** | **15.79 KB** | **15.69 KB** | **NODE_PROD**
ReactART-dev.js | -0.0% | -0.0% | 357.29 KB | 357.19 KB | 76.39 KB | 76.36 KB | FB_WWW_DEV
**ReactART-prod.js** | **-0.1%** | **-0.1%** | **153.2 KB** | **153.1 KB** | **26.42 KB** | **26.39 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | -0.0% | -0.0% | 363.36 KB | 363.25 KB | 80.07 KB | 80.05 KB | UMD_DEV
**react-test-renderer.production.min.js** | **-0.6%** | **-0.7%** | **49.03 KB** | **48.75 KB** | **15.16 KB** | **15.06 KB** | **UMD_PROD**
react-test-renderer.development.js | -0.0% | -0.0% | 359.49 KB | 359.38 KB | 79.11 KB | 79.08 KB | NODE_DEV
**react-test-renderer.production.min.js** | **-0.6%** | **-0.7%** | **48.74 KB** | **48.46 KB** | **15.02 KB** | **14.92 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | -0.0% | -0.0% | 363.95 KB | 363.85 KB | 77.9 KB | 77.87 KB | FB_WWW_DEV

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by gaearon** [comment]
@react-linter lol what. Something's messing up our reporting..

File: packages/react-reconciler/src/__tests__/ReactIncrementalErrorLogging-test.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    jest.resetModules();
+    React = require('react');
+    ReactNoop = require('react-noop-renderer');
+  });
+
+  it('should log errors that occur during the begin phase', () => {
+    // Manually override console.error() because our toWarn() matchers
+    // filter out precisely the messages we want to test for in this file.
+    const oldConsoleError = console.error;
+    console.error = jest.fn();
```
Comment (bvaughn) [inline packages/react-reconciler/src/__tests__/ReactIncrementalErrorLogging-test.js]: Why not do this override-and-restore stuff in `beforeEach`/`afterEach` ? Seems more "standard"

File: scripts/jest/matchers/toWarnDev.js [hunk 2]
```diff
@@ -53,7 +54,13 @@ const createMatcherFor = consoleMethod =>
       const isLikelyAComponentStack = message =>
         typeof message === 'string' && message.includes('\n    in ');
 
-      const consoleSpy = (format, ...args) => {
+      const createConsoleSpy = methodName => (format, ...args) => {
+        // Ignore uncaught errors reported by jsdom
+        // and React addendums because they're too noisy.
+        if (methodName === 'error' && shouldIgnoreConsoleError(format, args)) {
```
Comment (bvaughn) [inline scripts/jest/matchers/toWarnDev.js]: Totally unimportant but I'm just curious– was changing `consoleSpy` to `createConsoleSpy` actually unnecessary? Couldn't `methodName` have just used `consoleMethod`?

File: packages/react-reconciler/src/__tests__/ReactIncrementalErrorLogging-test.js (hunk 1 above)
Comment (bvaughn) [reply packages/react-reconciler/src/__tests__/ReactIncrementalErrorLogging-test.js]: Gotcha. I also don't feel strongly about this, was just curious.

**Pull request merged**